Combine deconstruction assignment and declaration, and support discards.

gafter opened the pull request:
- Combine deconstruction assignment and declaration, and support discards.
- Because tuples may contain declarations, adjust lambda disambiguation
  and adjust parsing of argument lists.
- Diagnose tuple element names on the left of a deconstruction.
- Add relational operators to disambiguating tokens in 7.5.4.2
- Remove `BoundVoid`

Fixes #15548 
Fixes #15549 

/cc @jcouv @AlekseyTs @VSadov @dotnet/roslyn-compiler

gafter commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -257,29 +246,40 @@ private void SetInferredTypes(ArrayBuilder<DeconstructionVariable> variables, Im
+                    {
+                        var pending = (BoundDiscardedExpression)expression;
+                        if ((object)pending.Type == null)
+                        {
```
Since the caller has already checked this, I will remove the `if` and add an assertion that the type is `null`.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -719,14 +744,20 @@ private DeconstructionVariable BindDeconstructionVariables(ExpressionSyntax node
 
+        private BoundDiscardedExpression BindDiscardedExpression(
+            DiscardedDesignationSyntax designation,
+            TypeSymbol declType)
```
`static`

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -703,6 +706,60 @@ private static bool IsThrowExpressionInProperContext(ThrowExpressionSyntax node)
+        }
+
+        // Bind a declaration variable where it isn't permitted.
+        private BoundExpression BindDeclarationVariables(TypeSymbol declType, VariableDesignationSyntax node, DiagnosticBag diagnostics)
```
Is this comment correct? From the body of the method, it's not clear that the declaration variable isn't permitted.

AlekseyTs replied:
>diagnostics [](start = 124, length = 11)

I am not sure why do we want to collect any additional errors besides the one already reported by the caller. It feels like that will only create more noise.

gafter replied:
We could drop the diagnostics from this call on the floor, but I believe they will be useful to the user when (1) the user refactors their code to use local declarations, or (2) we begin allowing this under more constrained circumstances. If we see complaints about cascaded diagnostics, we should consider refactoring. I'm fine with it as it stands.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -703,6 +706,60 @@ private static bool IsThrowExpressionInProperContext(ThrowExpressionSyntax node)
+                            builder.Add(BindDeclarationVariables(declType, n, diagnostics));
+                        }
+                        var subExpressions = builder.ToImmutableAndFree();
+                        var tupleType = (TypeSymbol)this.Compilation.CreateTupleTypeSymbol(subExpressions.Select<BoundExpression, ITypeSymbol>(e => e.Type).ToImmutableArray());
```
Consider using `TupleTypeSymbol` directly rather than `ITypeSymbol`, etc.:
```
TupleTypeSymbol.Create(
    null,
    subExpressions.SelectAsArray(e => e.Type),
    default(ImmutableArray<Location>),
    default(ImmutableArray<string>,
    Compilation);
```

AlekseyTs replied:
>var tupleType = (TypeSymbol)this.Compilation.CreateTupleTypeSymbol(subExpressions.Select<BoundExpression, ITypeSymbol>(e => e.Type).ToImmutableArray()); [](start = 24, length = 152)

Please use TupleTypeSymbol.Create API instead, see how BindTupleExpression does this.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -16,32 +16,21 @@ namespace Microsoft.CodeAnalysis.CSharp
+                // We only allow assignment-only or declaration-only deconstructions at this point.
+                // Issue https://github.com/dotnet/roslyn/issues/15050 tracks allowing mixed deconstructions.
+                // For now we give an error when you mix.
+                Error(diagnostics, ErrorCode.ERR_MixedDeconstructionUnsupported, left);
```
Please add a test for this case.

gafter replied:
There already is one.

---
In reply to: [89845635](https://github.com/dotnet/roslyn/pull/15548#discussion_r89845635) [](ancestors = 89845635)

cston commented on src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:
```diff
@@ -1426,14 +1425,19 @@ internal enum ErrorCode
+        ERR_MixedDeconstructionUnsupported = 8184,
+        ERR_DeclarationExpressionNotPermitted = 8185,
+        ERR_MustDeclareForeachIteration = 8186,
+        ERR_TupleElementNamesInDeconstruction = 8187,
```
Please add tests for each of these diagnostics.

gafter replied:
For ERR_MixedDeconstructionUnsupported: four tests in Emit\CodeGen\CodeGenDeconstructTests.cs, and two in Syntax\Parsing\DeconstructionTests.cs

for ERR_DeclarationExpressionNotPermitted: many tests in Emit\CodeGen\CodeGenTupleTest.cs, Semantic\Semantics\DeconstructionTests.cs, Semantic\Semantics\LocalFunctionTests.cs, Semantic\Semantics\OutVarTests.cs, and Semantic\Semantics\PatternSwitchTests.cs

for ERR_MustDeclareForeachIteration: two tests in Emit\CodeGen\CodeGenDeconstructTests.cs

for ERR_TupleElementNamesInDeconstruction: two tests in Emit\CodeGen\CodeGenDeconstructTests.cs

Are you saying that you don't think these tests are sufficient?

---
In reply to: [89845938](https://github.com/dotnet/roslyn/pull/15548#discussion_r89845938) [](ancestors = 89845938)

cston replied:
Sorry, I missed those tests since the webpage did not include those in the diff.

cston commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -255,7 +277,7 @@ private BoundForEachStatement BindForEachPartsWorker(DiagnosticBag diagnostics,
 
             // Skip the conversion checks and array/enumerator differentiation if we know we have an error (except local name conflicts).
-            if (hasErrors)
+            if (hasErrors || _syntax.HasErrors)
```
Is the check for `_syntax.HasErrors` necessary? What if the error is something simple such as missing `in` token?

gafter replied:
I'm fine with suppressing semantic errors in the face of syntax errors, given that we've bound all the pieces of the syntax needed for a good IDE/SemanticModel experience.

---
In reply to: [89846888](https://github.com/dotnet/roslyn/pull/15548#discussion_r89846888) [](ancestors = 89846888)

gafter replied:
I'll change the syntax errors check to check for syntax errors only on the declaration part.

cston commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -8000,33 +8014,59 @@ private CommonForEachStatementSyntax ParseForEachStatement()
-            return (deconstruction == null)
-                ? (CommonForEachStatementSyntax)_syntaxFactory.ForEachStatement(@foreach, openParen, type, name, @in, expression, closeParen, statement)
-                : (CommonForEachStatementSyntax)_syntaxFactory.ForEachVariableStatement(@foreach, openParen, deconstruction, @in, expression, closeParen, statement);
+        private bool IsValidForeachVariable(ExpressionSyntax variable)
```
>private [](start = 8, length = 7)

`static`

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs:
```diff
@@ -102,6 +102,132 @@ private PatternSyntax ParsePattern()
+        // differently, as we check for the comma before concluding that the identifier should cause a
+        // disambiguation. For example, for the input `(A < B , C > D)`, we treat this as a tuple with
+        // two elements, because if we considered the `A<B,C>` to be a type, it wouldn't be a tuple at
+        // all. Since we don't have such as thing as a one-element tuple (even for deconstruction), the
```
>such as thing [](start = 36, length = 13)

Typo "such **as** thing" vs "such **a** thing"?

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs:
```diff
@@ -102,6 +102,132 @@ private PatternSyntax ParsePattern()
+                }
+
+                // check for a designation
+                if (!ScanDesignation(permitTupleDesignation && (lastTokenOfType.IsVar() || IsPredefinedType(lastTokenOfType.Kind))))
```
>lastTokenOfType.IsVar() [](start = 64, length = 23)

Will this condition do the right thing for a type name "A.var"?

gafter replied:
Probably not, as we'll want `A.var(x, y)` to be an invocation expression. I'll revise it.

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs:
```diff
@@ -102,6 +102,132 @@ private PatternSyntax ParsePattern()
+                    return false;
+                case SyntaxKind.IdentifierToken:
+                    var result = this.IsTrueIdentifier();
+                    this.EatToken();
```
>this.EatToken(); [](start = 20, length = 16)

It doesn't feel appropriate to eat the token when we are about to return false.

gafter replied:
When we return false, the caller unwinds the trial parse, so it doesn't matter.

AlekseyTs replied:
I think it would be cleaner and more intuitive if this function (and the like) wouldn't consume tokens  it doesn't accept.

gafter replied:
That isn't possible given the way we scan forward and discard tokens before making a decision. Even if we don't eat the very last token, we will have eaten preceding tokens that we don't accept. Avoiding eating the very last token would be extra work that serves no purpose.

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs:
```diff
@@ -102,6 +102,132 @@ private PatternSyntax ParsePattern()
+                        switch (this.CurrentToken.Kind)
+                        {
+                            case SyntaxKind.CloseParenToken:
+                                this.EatToken();
```
>this.EatToken(); [](start = 32, length = 16)

It doesn't feel appropriate to eat the token when we are about to return false.

gafter replied:
As with other scan methods, it is unspecified what tokens have been eaten when it returns false.

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5346,6 +5343,13 @@ private bool IsTrueIdentifier()
             return false;
         }
 
+        private bool IsTrueIdentifier(SyntaxToken token)
```
>IsTrueIdentifier [](start = 21, length = 16)

Should the method above delegate its work to this method? Otherwise, the methods have the same name, but appear to use different criteria.

gafter replied:
Good idea, but not quite so simple to do, as there are a series of helper methods for the previous method intended to help it disambiguate, implicitly based on the current token in declarations; it peeks forward using `PeekToken(int)` and even does some trial forward parsing.

This method is for use in executable code (not declarations), so it has a much simpler implementation that depends only on the token passed in and `this.IsInQuery`. I'll add some clarifying doc comments.

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6450,11 +6463,34 @@ private enum ParseTypeMode
+            var type = this.ParseUnderlyingType(parentIsParameter: mode == ParseTypeMode.Parameter, options: nameOptions);
+
+            if (this.CurrentToken.Kind == SyntaxKind.QuestionToken &&
+                (mode != ParseTypeMode.AfterIsOrCase || !IsTrueIdentifier(this.PeekToken(1))))
```
>(mode != ParseTypeMode.AfterIsOrCase || !IsTrueIdentifier(this.PeekToken(1)))) [](start = 16, length = 78)

Consider adding a comment explaining the rational behind this new condition.

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -6935,8 +6986,8 @@ private bool IsPossibleLocalDeclarationStatement(bool allowAnyExpression)
-            if (tk == SyntaxKind.RefKeyword ||
-                (SyntaxFacts.IsPredefinedType(tk) && this.PeekToken(1).Kind != SyntaxKind.DotToken) || IsDeclarationModifier(tk))
+            if (tk == SyntaxKind.RefKeyword || IsDeclarationModifier(tk) ||
+                (SyntaxFacts.IsPredefinedType(tk) && this.PeekToken(1).Kind != SyntaxKind.DotToken && this.PeekToken(1).Kind != SyntaxKind.OpenParenToken))
```
>this.PeekToken(1).Kind != SyntaxKind.OpenParenToken [](start = 102, length = 51)

Consider adding a clarifying comment for this condition.

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -7525,28 +7571,14 @@ private StatementSyntax ParseEmbeddedStatement(bool complexCheck)
-                        var assignment = (AssignmentExpressionSyntax)expression;
-                        if (assignment.Left.IsDeconstructionDeclarationLeft())
-                        {
-                            statement = this.AddError(statement, ErrorCode.ERR_BadEmbeddedStmt);
```
>statement = this.AddError(statement, ErrorCode.ERR_BadEmbeddedStmt); [](start = 28, length = 68)

It is not clear why this error is no longer reported. Was there an explicit LDM decision?

gafter replied:
Yes, the LDM explicitly decided to remove the language concept of a deconstruction declaration statement, which is where this restriction arises. We will discuss tomorrow if the LDM wants to add back any of these restrictions in C# 7, including the ones among the first changes in `Binder_Deconstruct.cs` (lines 26-57 in this PR); as of the last LDM they wanted to remove all those restrictions. But when we do we will need additional tests.

AlekseyTs replied:
The way I read the response, and based on my recollection of the LDM discussion, the answer to my question "Was there an explicit LDM decision?" is No. Meaning the LDM did not make an explicit decision about this error. So, let's hold on on removing it until the decision is made.

gafter replied:
You misread the response. We did discuss the error, I raised the point that it is no longer an error if it is not a declaration when in the position of a controlled statement, and that was OK with Mads and others. If the LDM explicitly decides to change their mind and re-introduce an error here, I will do that.

gafter replied:
The LDM met yesterday and reconfirmed this. From the raw LDM notes:

> Declarations in embedded statements?
> =====================================
> 
> Let's not add errors for deconstruction situations. In principle we probably want to remove even the existing limitation, but it's work we won't push for in C# 7.0.

AlekseyTs commented:
>             Debug.Assert(this.CurrentToken.Kind == SyntaxKind.ForKeyword || this.CurrentToken.Kind == SyntaxKind.ForEachKeyword);

Should this condition be now removed?

---
Refers to: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:7832 in 1003470. [](commit_id = 1003470c5f0de8925d3e1f465a59f549e2592eb2, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -7812,28 +7844,22 @@ private StatementSyntax ParseForOrForEachStatement()
             try
             {
-                if (this.CurrentToken.Kind == SyntaxKind.ForKeyword)
+                Debug.Assert(this.CurrentToken.Kind == SyntaxKind.ForKeyword);
```
>Debug.Assert(this.CurrentToken.Kind == SyntaxKind.ForKeyword); [](start = 16, length = 62)

It looks like we already asserted this at the beginning of this function.

cston commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -9898,8 +9804,17 @@ internal BracketedArgumentListSyntax ParseBracketedArgumentList()
+                    this.CurrentToken.Kind == SyntaxKind.CloseBracketToken)
+                {
+                    // convert `]` into `)` or vice versa for error recovery
+                    closeToken = this.ConsumeOneToken(closeKind);
```
>this.ConsumeOneToken(closeKind) [](start = 33, length = 31)

Why consume the ']' or ')' if it's not the close kind expected? Is it more likely the user closed with the wrong kind or that the expected kind is missing?

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -8482,200 +8514,7 @@ private StatementSyntax ParseLocalDeclarationStatement()
-        }
-
-        private VariableDesignationSyntax ParseDeconstructionDesignation(bool topLevel = false)
+        private VariableDesignationSyntax ParseDesignation(bool topLevel = false)
```
>bool topLevel = false [](start = 59, length = 21)

It looks like this parameter is not used.

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -9597,6 +9416,12 @@ private ExpressionSyntax ParseTerm(Precedence precedence)
                         }
+                        else if (this.IsPossibleDeconstructionLeft(precedence))
+                        {
+                            TypeSyntax type = this.ParseType();
```
>TypeSyntax type = this.ParseType(); [](start = 28, length = 35)

It looks like the code in this block is a dupe of a new code in ParseSubExpressionCore. Consider extracting into a method and reusing.

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -9839,7 +9732,18 @@ internal BracketedArgumentListSyntax ParseBracketedArgumentList()
             bool isIndexer = openKind == SyntaxKind.OpenBracketToken;
-            var open = this.EatToken(openKind);
+
+            if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken ||
```
>if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken || [](start = 12, length = 58)

Consider adding an assert that
- ```openKind``` is either SyntaxKind.OpenParenToken or SyntaxKind.OpenBracketToken
- ```closeKind``` is either SyntaxKind.CloseParenToken or SyntaxKind.CloseBracketToken

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs:
```diff
@@ -102,6 +102,132 @@ private PatternSyntax ParsePattern()
+            {
+                TypeSyntax type = this.ParseType(mode);
+                var designation = ParseDesignation();
+                if (feature != 0)
```
>feature != 0 [](start = 20, length = 12)

Consider adding a special enum member rather than using a magic number 0.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs:
```diff
@@ -172,21 +166,27 @@ static object Test1(out int x)
                 );
 
-            Assert.False(compilation.SyntaxTrees.Single().GetRoot().DescendantNodes().OfType<DeclarationExpressionSyntax>().Any());
+            Assert.Equal(2, compilation.SyntaxTrees.Single().GetRoot().DescendantNodes().OfType<DeclarationExpressionSyntax>().Count());
```
>Assert.Equal(2, compilation.SyntaxTrees.Single().GetRoot().DescendantNodes().OfType<DeclarationExpressionSyntax>().Count()); [](start = 12, length = 124)

Please test SemanticModel behavior around these expressions, their designators, and identifier names x1 and x2.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs:
```diff
@@ -8991,9 +8994,6 @@ static bool TakeOutParam(object y, out object x)
             var compilation = CreateCompilationWithMscorlib45(source, references: new[] { ValueTupleRef, SystemRuntimeFacadeRef },
                                                               options: TestOptions.DebugExe, parseOptions: TestOptions.Regular);
             compilation.VerifyDiagnostics(
-                // (11,13): error CS1023: Embedded statement cannot be a declaration or labeled statement
```
>// (11,13): error CS1023: Embedded statement cannot be a declaration or labeled statement [](start = 16, length = 89)

I think we should keep this error, unless all targets are discards.

gafter replied:
If the LDM decides to reverse course and change the specification back to a declaration, we can add the error back.

AlekseyTs replied:
I don't think the two are necessarily connected, we should discuss the error on its own. Let's do this tomorrow then.

gafter replied:
The LDM has confirmed that the removal of this error is intended.

AlekseyTs commented:
>   <data name="ERR_MixedDeconstructionDisallowed" xml:space="preserve">

It looks like this error id has been removed.

---
Refers to: src/Compilers/CSharp/Portable/CSharpResources.resx:2108 in 1003470. [](commit_id = 1003470c5f0de8925d3e1f465a59f549e2592eb2, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -703,6 +706,60 @@ private static bool IsThrowExpressionInProperContext(ThrowExpressionSyntax node)
+                case SyntaxKind.DiscardedDesignation:
+                    {
+                        var discarded = (DiscardedDesignationSyntax)node;
+                        return BindDiscardedExpression(discarded, declType);
```
>declType [](start = 66, length = 8)

Should we pass an error type if ```declType``` is null?

gafter replied:
Yes; I'll move that to entry to this method.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -703,6 +706,60 @@ private static bool IsThrowExpressionInProperContext(ThrowExpressionSyntax node)
+                case SyntaxKind.ParenthesizedVariableDesignation:
+                    {
+                        var tuple = (ParenthesizedVariableDesignationSyntax)node;
+                        var builder = ArrayBuilder<BoundExpression>.GetInstance();
```
>GetInstance(); [](start = 68, length = 14)

Can we predict the number of elements we will need?

jcouv commented on src/Compilers/CSharp/Portable/Syntax/InternalSyntax/TypeSyntax.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    internal abstract partial class TypeSyntax
+    {
+        public bool IsVar
```
There is a similar implementation on the public `TypeSyntax` (in `D:\repos\roslyn\src\Compilers\CSharp\Portable\Syntax\TypeSyntax.cs`). Maybe it could be factored to call this one or be removed completely?

jcouv replied:
Also, there is an `IsVar` extension method in `src\Compilers\CSharp\Portable\Syntax\SyntaxFacts.cs`. Could it be used instead?

---
In reply to: [90152557](https://github.com/dotnet/roslyn/pull/15548#discussion_r90152557) [](ancestors = 90152557)

gafter replied:
The internal and public TypeSyntax are different (unrelated) types, so the extension methods are necessarily different.

gafter commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -8000,33 +8014,59 @@ private CommonForEachStatementSyntax ParseForEachStatement()
+            var @in = this.EatToken(SyntaxKind.InKeyword, ErrorCode.ERR_InExpected);
+            if (!IsValidForeachVariable(variable))
+            {
+                @in = this.AddError(@in, ErrorCode.ERR_BadForeachDecl);
```
Perhaps, but this models the diagnostics we produce currently. I'd prefer to do improvements in a separate PR.

AlekseyTs replied:
> I'd prefer to do improvements in a separate PR.

Please open an issue to track the work then.

gafter replied:
I do not believe there is any work here worth doing. If you feel otherwise please open an issue.

AlekseyTs replied:
I am confused. First you say: "I'd prefer to do improvements in a separate PR." Then you say: "I do not believe there is any work here worth doing." Why do you think placing errors at the right location isn't worth doing?

AlekseyTs replied:
How much cost is behind this?

gafter replied:
I believe that the error is at the right location, but I'm open to an investigation about moving it.

---
In reply to: [90168970](https://github.com/dotnet/roslyn/pull/15548#discussion_r90168970) [](ancestors = 90168970)

gafter replied:
The easiest way to find out would be to do the work.

---
In reply to: [90169065](https://github.com/dotnet/roslyn/pull/15548#discussion_r90169065) [](ancestors = 90169065)

gafter replied:
I've filed an issue https://github.com/dotnet/roslyn/issues/15632 for this suggestion, though I don't believe the suggestion would be an improvement.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -233,16 +236,35 @@ private BoundForEachStatement BindForEachPartsWorker(DiagnosticBag diagnostics,
-                                                                                rightPlaceholder: valuePlaceholder);
-
-                        deconstructStep = new BoundForEachDeconstructStep(variables, deconstruction, valuePlaceholder);
+                        if (variables.Kind() == SyntaxKind.TupleExpression || variables.Kind() == SyntaxKind.DeclarationExpression)
```
>variables.Kind() == SyntaxKind.TupleExpression || variables.Kind() == SyntaxKind.DeclarationExpressio [](start = 28, length = 101)

Should we use IsDeconstructionLeft helper instead?

gafter commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -682,7 +685,7 @@ internal override SyntaxNode ForbiddenZone
                     {
                         case SyntaxKind.SimpleAssignmentExpression:
-                            return ((AssignmentExpressionSyntax)_deconstruction).Right;
+                            return _deconstruction;
```
Yes; see `CodeGenDeconstructTests.cs`, test `DeclareAndUseLocalInDeconstruction`

AlekseyTs commented:
>         private static bool ContainsDeclarations(ExpressionSyntax expression)

It looks like this method is no longer used.

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:63 in 1003470. [](commit_id = 1003470c5f0de8925d3e1f465a59f549e2592eb2, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -633,22 +659,44 @@ private static void FlattenDeconstructVariables(ArrayBuilder<DeconstructionVaria
+        private DeconstructionVariable BindDeconstructionVariables(
+            ExpressionSyntax node,
+            DiagnosticBag diagnostics,
+            ref CSharpSyntaxNode declaration,
```
>ref CSharpSyntaxNode declaration, [](start = 12, length = 33)

This and the next parameter deserve a comment or should be renamed to use more descriptive names.

AlekseyTs replied:
>CSharpSyntaxNode [](start = 16, length = 16)

It looks like we can use more specific type for this parameter - DeclarationExpressionSyntax.

gafter replied:
The caller doesn't need that, but I'm OK with using a more specific type here.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:
```diff
@@ -5056,23 +4989,9 @@ static void Main()
     }
 }
 ";
+            // This is permitted, as long as it doesn't use an expression lvalue in the deconstruction.
```
>// This is permitted, as long as it doesn't use an expression lvalue in the deconstruction. [](start = 12, length = 91)

Please add a test that verifies that the behavior of the compiled code is correct. Please test semantic model.

gafter replied:
Julien is working on semantic model support now. I'll do a runtime test.

---
In reply to: [90168304](https://github.com/dotnet/roslyn/pull/15548#discussion_r90168304) [](ancestors = 90168304)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:
```diff
@@ -1151,15 +1151,7 @@ public void Deconstruct(out int a, out int b)
-                // (6,19): error CS0825: The contextual keyword 'var' may only appear within a local variable declaration or in script code
-                //         var z = ((var x, int y) = new C());
-                Diagnostic(ErrorCode.ERR_TypeVarNotFound, "var").WithLocation(6, 19)
+                // this is permitted now, as it is just a deconstruction assignment expression
```
>// this is permitted now, as it is just a deconstruction assignment expression [](start = 16, length = 78)

Please add a test that verifies that the behavior of the compiled code is correct. Test SemanticModel as well.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:
```diff
@@ -2735,25 +2691,8 @@ static void Main()
 ";
 
             var comp = CreateCompilationWithMscorlib(source, references: s_valueTupleRefs);
+            // this is permitted now, as it is just an assignment expression
```
>// this is permitted now, as it is just an assignment expression [](start = 12, length = 64)

Please add a test that verifies that the behavior of the compiled code is correct. Test SemanticModel as well.

AlekseyTs commented:
>         foreach (_ in M())

Please test SemanticModel behavior for this expression.

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:5418 in 1003470. [](commit_id = 1003470c5f0de8925d3e1f465a59f549e2592eb2, deletion_comment = False)

gafter commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -233,16 +236,35 @@ private BoundForEachStatement BindForEachPartsWorker(DiagnosticBag diagnostics,
+                        else if (!node.HasErrors)
+                        {
+                            // error: must declare foreach loop iteration variables.
+                            Error(diagnostics, ErrorCode.ERR_MustDeclareForeachIteration, variables);
```
No; we never handled badly broken code in this context before, and we don't have any tests to prove that we didn't and still don't. I'll open an issue for this suggestion.

gafter replied:
Added #15601 for this suggestion.

AlekseyTs replied:
This change fails do bind a portion of the syntax, this is introduced by this change, before ```variables``` were always bound.

---
In reply to: [90168965](https://github.com/dotnet/roslyn/pull/15548#discussion_r90168965) [](ancestors = 90168965)

gafter replied:
Also #15606

---
In reply to: [90169155](https://github.com/dotnet/roslyn/pull/15548#discussion_r90169155) [](ancestors = 90169155)

gafter replied:
Added a SemanticModel test for this, and two issues for further follow-up.

---
In reply to: [90164938](https://github.com/dotnet/roslyn/pull/15548#discussion_r90164938) [](ancestors = 90164938)

AlekseyTs replied:
Could you please point me to the test that you added?

---
In reply to: [90179660](https://github.com/dotnet/roslyn/pull/15548#discussion_r90179660) [](ancestors = 90179660,90164938)

AlekseyTs replied:
We do bind "badly broken" code. For example, we bind and preserve nodes for arguments in a variable declarator. So far, for an executable code, our policy was to bind and preserve a all expressions that are part of a syntax.  

---
In reply to: [90174159](https://github.com/dotnet/roslyn/pull/15548#discussion_r90174159) [](ancestors = 90174159,90168965)

AlekseyTs replied:
We do bind "badly broken" code. For example, we bind and preserve nodes for arguments in a variable declarator. So far, for an executable code, our policy was to bind and preserve all expressions that are part of a syntax.  

---
In reply to: [91386197](https://github.com/dotnet/roslyn/pull/15548#discussion_r91386197) [](ancestors = 91386197,90179660,90164938)

gafter replied:
`CodeGenDeconstructTests.ShortDiscardDisallowedInForeach`  is the only way this can occur (i.e. a simple name `_`). There is no semantic model test as the semantic model stuff for wildcards is an open issue and that is noted in the test.

---
In reply to: [91419535](https://github.com/dotnet/roslyn/pull/15548#discussion_r91419535) [](ancestors = 91419535,91386197,90179660,90164938)

gafter replied:
I'll open a separate issue for this. As it stands there are no behavioral symptoms.

---
In reply to: [91419166](https://github.com/dotnet/roslyn/pull/15548#discussion_r91419166) [](ancestors = 91419166,90174159,90168965)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -11022,12 +11022,15 @@ static void Main()
                 //         System.Console.WriteLine(nameof((int a, int b)));
-                Diagnostic(ErrorCode.ERR_SyntaxError, ")").WithArguments("=>", ")").WithLocation(7, 55),
-                // (7,55): error CS1525: Invalid expression term ')'
+                Diagnostic(ErrorCode.ERR_DeclarationExpressionNotPermitted, "int a").WithLocation(7, 42),
```
>ERR_DeclarationExpressionNotPermitted [](start = 37, length = 37)

Please test SemanticModel behavior for the syntax nodes that now become the part of the syntax tree: tuple expression, declaration expression, designators, etc.

gafter replied:
I don't believe we've ever tested semantic model for declarations appearing in nameof(). I doubt we bind them into an expression that would make the SemanticModel work. I'll open an issue: #15604

---
In reply to: [90169278](https://github.com/dotnet/roslyn/pull/15548#discussion_r90169278) [](ancestors = 90169278)

AlekseyTs commented:
>         if (o is (int a, int b)) { }

Should we parse this as a tuple type instead?

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:13261 in 1003470. [](commit_id = 1003470c5f0de8925d3e1f465a59f549e2592eb2, deletion_comment = False)

AlekseyTs commented:
>         (int x1, string x2);

Please test SemanticModel for the nodes affected by the change. 

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs:1430 in 1003470. [](commit_id = 1003470c5f0de8925d3e1f465a59f549e2592eb2, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs:
```diff
@@ -2620,5 +2634,91 @@ static void Main()
+
+            var comp = CreateCompilationWithMscorlib(source, references: new[] { ValueTupleRef, SystemRuntimeFacadeRef });
+            comp.VerifyDiagnostics(
+                // (7,18): error CS8185: A declaration is not allowed in this context.
```
>// (7,18): error CS8185: A declaration is not allowed in this context. [](start = 16, length = 70)

It is not clear why this error is reported. Could you explain, please.

gafter replied:
Because per LDM we need a separate decision (from Jared) when  (i.e. either C# 7 or 7.1 or 8) to permit the declarations in deconstructions in contexts other than an expression statement containing the deconstruction assignment and an expression among the for-loop initializers. I believe he is likely to make that decision tomorrow.

---
In reply to: [90169794](https://github.com/dotnet/roslyn/pull/15548#discussion_r90169794) [](ancestors = 90169794)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs:
```diff
@@ -2620,5 +2634,91 @@ static void Main()
         }
+
+        [Fact]
+        public void MixedDeconstruction_01()
```
>MixedDeconstruction_01 [](start = 20, length = 22)

Please test SemanticModel for the all new unit-tests below.

gafter replied:
Sure, but I will have to move them to CodeGenDeconstructTests where the infrastructure exists for that.

---
In reply to: [90169982](https://github.com/dotnet/roslyn/pull/15548#discussion_r90169982) [](ancestors = 90169982)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs:
```diff
@@ -2620,5 +2634,91 @@ static void Main()
+
+            var comp = CreateCompilationWithMscorlib(source, references: new[] { ValueTupleRef, SystemRuntimeFacadeRef });
+            comp.VerifyDiagnostics(
+                // (7,19): error CS8185: A declaration is not allowed in this context.
```
>// (7,19): error CS8185: A declaration is not allowed in this context. [](start = 16, length = 70)

It is not clear why this error is reported. Could you explain, please?

gafter replied:
Yes, it is for Jared to decide tomorrow whether deconstruction declarations are to be permitted anywhere but as the expression of an expression statement and as one of the expressions in the expression list of a for initializer.

---
In reply to: [90170034](https://github.com/dotnet/roslyn/pull/15548#discussion_r90170034) [](ancestors = 90170034)

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs:
```diff
@@ -254,12 +254,13 @@ public static int Main()
-    // (7,22): error CS0230: Type and identifier are both required in a foreach statement
-    //         foreach (int in myarray)   // CS0230
-    Diagnostic(ErrorCode.ERR_BadForeachDecl, "in"));
+                // (7,18): error CS1525: Invalid expression term 'int'
```
>// (7,18): error CS1525: Invalid expression term 'int' [](start = 16, length = 54)

It feels like the new error is confusing. We know that ```int``` is a type. Why do we try parse it as an expression?

gafter replied:
Because in the absence of a simple declaration before the `in`, we parse it as an expression, such as a tuple lvalue or a tuple declaration expression.

---
In reply to: [90170673](https://github.com/dotnet/roslyn/pull/15548#discussion_r90170673) [](ancestors = 90170673)

AlekseyTs commented on src/EditorFeatures/CSharpTest/Diagnostics/GenerateMethod/GenerateMethodTests.cs:
```diff
@@ -7246,7 +7246,7 @@ public void M1()
         }
 
-        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/15508"), Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
```
>Skip = "https://github.com/dotnet/roslyn/issues/15508 [](start = 14, length = 53)

It feels wrong that parser now produces single type tuple types or tuples. I believe language grammar doesn't allow such syntax. I think it should add a missing comma token and missing type/expression for the second element, this is what we do for other similar situations.

gafter replied:
My mistake. The parser does not do that. The parser produces `(int, int)` as an expression, and the IDE code chokes on that. I'll fix the bug description.

---
In reply to: [90171673](https://github.com/dotnet/roslyn/pull/15548#discussion_r90171673) [](ancestors = 90171673)

AlekseyTs replied:
>Skip = "https://github.com/dotnet/roslyn/issues/15508" [](start = 14, length = 54)

It is fine to merge with skipped test into a feature branch, but I think we should come up with solution or implement a workaround that prevents the crash before we merge this to master.

jaredpar replied:
Agree. This is an ask mode change to RC.  We can't introduce new IDE regressions here without explicit IDE buy off.  Otherwise this will not pass the ask mode review.

gafter replied:
That (not crashing) was the reason for the addition of "return false" in "CSharpTypeInferenceService.TypeInferrer.cs".

I'll adjust this test so it reflects the "broken" but non-crashing behavior.

---
In reply to: [91455957](https://github.com/dotnet/roslyn/pull/15548#discussion_r91455957) [](ancestors = 91455957)

AlekseyTs commented:
Done with review pass.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ExpressionVariableFinder.cs:
```diff
@@ -260,7 +260,7 @@ public override void VisitBinaryExpression(BinaryExpressionSyntax node)
         {
             var argumentSyntax = (ArgumentSyntax)node?.Parent;
-            var argumentListSyntax = (BaseArgumentListSyntax)argumentSyntax?.Parent;
+            var argumentListSyntax = argumentSyntax.FirstAncestorOrSelf<BaseArgumentListSyntax>();
```
I think we shouldn't be creating Out Variables for such cases, they should be created only for the Out arguments.   

---
In reply to: [90172205](https://github.com/dotnet/roslyn/pull/15548#discussion_r90172205) [](ancestors = 90172205,90163033)

gafter replied:
Yes, it was wrong now that declaration expressions can appear in new and novel contexts. However, perhaps it would be better to replace the (incorrect) cast with an `as` operator.

gafter replied:
That code works fine for this situation. Custom code for this error situation would do the same thing.

---
In reply to: [90172670](https://github.com/dotnet/roslyn/pull/15548#discussion_r90172670) [](ancestors = 90172670,90172205,90163033)

gafter commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -233,16 +236,35 @@ private BoundForEachStatement BindForEachPartsWorker(DiagnosticBag diagnostics,
+
+                            deconstructStep = new BoundForEachDeconstructStep(variables, deconstruction, valuePlaceholder);
+                        }
+                        else if (!node.HasErrors)
```
That will be `MixedDeconstruction_05` in `CodeGenDeconstructTests.cs` in the next iteration.

---
In reply to: [90174311](https://github.com/dotnet/roslyn/pull/15548#discussion_r90174311) [](ancestors = 90174311)

gafter replied:
`CodeGenDeconstructTests.ShortDiscardDisallowedInForeach`  is the only way this can occur (i.e. a simple name `_`).

---
In reply to: [91418538](https://github.com/dotnet/roslyn/pull/15548#discussion_r91418538) [](ancestors = 91418538,90178336,90174311)

gafter commented:
>         foreach (_ in M())

That is a work item that Julien has in progress.

---
In reply to: [263782887](https://github.com/dotnet/roslyn/pull/15548#issuecomment-263782887) [](ancestors = 263782887)

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:5418 in 1003470. [](commit_id = 1003470c5f0de8925d3e1f465a59f549e2592eb2, deletion_comment = False)

gafter commented:
>         if (o is (int a, int b)) { }

No. The LDM explicitly took this decision quite some time ago.

---
In reply to: [263784217](https://github.com/dotnet/roslyn/pull/15548#issuecomment-263784217) [](ancestors = 263784217)

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:13261 in 1003470. [](commit_id = 1003470c5f0de8925d3e1f465a59f549e2592eb2, deletion_comment = False)

gafter commented:
>         (int x1, string x2);

OK, will move it to CodeGenDeconstructTests and do so.

---
In reply to: [263784556](https://github.com/dotnet/roslyn/pull/15548#issuecomment-263784556) [](ancestors = 263784556)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs:1430 in 1003470. [](commit_id = 1003470c5f0de8925d3e1f465a59f549e2592eb2, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -16,32 +16,47 @@ namespace Microsoft.CodeAnalysis.CSharp
         {
             var left = node.Left;
             var right = node.Right;
+            CSharpSyntaxNode declaration = null;
```
Since `declaration` and `expression` must be either `null` or `node`, could they just be Boolean values (`isDeclaration` and `isExpression`)?

gafter replied:
No, they are used as a location for a diagnostic below.

---
In reply to: [90290815](https://github.com/dotnet/roslyn/pull/15548#discussion_r90290815) [](ancestors = 90290815)

jcouv commented:
>         foreach (_ in M())

For other tests, I tried to verify anything that is already functioning. 
If the semantic model doesn't work for this case yet, please put a comment referring to [https://github.com/dotnet/roslyn/issues/15450](https://github.com/dotnet/roslyn/issues/15450)


---
In reply to: [263801699](https://github.com/dotnet/roslyn/pull/15548#issuecomment-263801699) [](ancestors = 263801699,263782887)

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:5418 in 1003470. [](commit_id = 1003470c5f0de8925d3e1f465a59f549e2592eb2, deletion_comment = False)

jcouv commented:
>         public void DeclarationCannotBeEmbedded()

Since this scenario is now allowed, test should be renamed to `DeconstructionCanBeEmbedded`

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/DeconstructionTests.cs:2413 in ac9c1a3. [](commit_id = ac9c1a39a21de907a9b4bf8f91295e11ec8f04ea, deletion_comment = False)

jcouv commented:
>                 Diagnostic(ErrorCode.ERR_VarInvocationLvalueReserved, "var (x1, x2)").WithLocation(6, 19),

Just to confirm, this is still expected (parser blocking `var (x1, x2)` in out var scenario)?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs:83 in ac9c1a3. [](commit_id = ac9c1a39a21de907a9b4bf8f91295e11ec8f04ea, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationExpressionTests.cs:
```diff
@@ -0,0 +1,1186 @@
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Xunit;
```
Nit: sorting

gafter commented:
>                 Diagnostic(ErrorCode.ERR_VarInvocationLvalueReserved, "var (x1, x2)").WithLocation(6, 19),

In this context it is simply an invocation expression of a kind that is contextually forbidden. The parser won't produce a declaration expression with a tuple designator in this context.

---
In reply to: [263964319](https://github.com/dotnet/roslyn/pull/15548#issuecomment-263964319) [](ancestors = 263964319)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs:83 in ac9c1a3. [](commit_id = ac9c1a39a21de907a9b4bf8f91295e11ec8f04ea, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationExpressionTests.cs:
```diff
@@ -0,0 +1,1186 @@
+        [Fact]
+        public void NewForeach_07()
+        {
+            // PROTOTYPE(wildcards): SHOULD BE some kind of *semantic* error here
```
If you want this PR to target `master` branch (for RC3), don't forget to replace the PROTOTYPE markers with issues.

jcouv commented on src/Compilers/CSharp/Test/Syntax/Parsing/DeconstructionTests.cs:
```diff
@@ -2071,7 +2071,6 @@ struct ValueTuple<T1, T2>
         public ValueTuple(T1 item1, T2 item2) { this.Item1 = item1; this.Item2 = item2; }
     }
 }";
-            // the duplicate reporting of CS8136 below is due to open issue https://github.com/dotnet/roslyn/issues/12905
```
>https://github.com/dotnet/roslyn/issues/12905 [](start = 76, length = 45)

Should that issue be resolved?

gafter replied:
Perhaps. It needs to be tested in other contexts as well. I suspect there is an issue in `foreach`.

---
In reply to: [90307970](https://github.com/dotnet/roslyn/pull/15548#discussion_r90307970) [](ancestors = 90307970)

gafter commented:
@AlekseyTs @jcouv @cston All comments have been addressed. I've also renamed "Discarded" to "Discard" per yesterday's LDM. Please re-review.

jcouv commented on src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs:
```diff
@@ -102,6 +102,148 @@ private PatternSyntax ParsePattern()
+                    while (true)
+                    {
+                        this.EatToken(); // consume the `(` or `,`
+                        if (!ScanDesignation(true))
```
Consider `ScanDesignation(permitTuple)` or named parameter.

jcouv commented on src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs:
```diff
@@ -474,6 +474,22 @@ protected SyntaxToken EatToken(SyntaxKind kind)
             return CreateMissingToken(kind, this.CurrentToken.Kind, reportError: true);
         }
 
+        // Consume a token if it is the right kind. Otherwise skip a token and replace it with one of the correct kind.
```
Suggestion:
`// Eat and return the current token, but wrapped it as a missing token if it is not the expected kind.`

jcouv replied:
The meaning of consume vs. eat is not obvious. Maybe name `ForceEatToken`?

gafter commented:
@AlekseyTs @jcouv @cston All comments have been addressed. Please re-review.

AlekseyTs commented:
@gafter 
> All comments have been addressed. Please re-review.

I'll try to do this today.

jcouv commented on src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/DeclarationTests.cs:
```diff
@@ -228,8 +231,12 @@ .maxstack  4
-}");
+  IL_003e:  ldloc.1
+  IL_003f:  ldloc.2
+  IL_0040:  newobj     ""System.ValueTuple<int, string>..ctor(int, string)""
```
I didn't understand what caused this change. The emitter is only supposed to emit the construction of the ValueTuple result if it is `used`.

gafter replied:
This is for the result of `CompileExpression`... it is the result of the expression. The expression compiler always produces the result of the expression for the consumer to use. This change is caused by the fact that the result of the deconstruction is now a tuple instead of void.

jcouv commented on src/EditorFeatures/CSharpTest/Formatting/Indentation/SmartIndenterTests.cs:
```diff
@@ -2570,7 +2570,7 @@ class C
         }
 
-        [WpfFact, Trait(Traits.Feature, Traits.Features.SmartIndent)]
+        [WpfFact(Skip = "PROTOYPE(wildcard): parse tree changed invalidating this test"), Trait(Traits.Feature, Traits.Features.SmartIndent)]
```
Remove PROTOTYPE if you want to re-target to master branch.

gafter replied:
Strange that VS's "find all" doesn't find this string... Oh, PROTOTYPE was misspelled.

AlekseyTs commented:
> I'll try to do this today.

Sorry. Cannot take a look today. Will look at the change tomorrow morning/early afternoon.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ExpressionVariableFinder.cs:
```diff
@@ -260,7 +260,7 @@ public override void VisitBinaryExpression(BinaryExpressionSyntax node)
         public override void VisitDeclarationExpression(DeclarationExpressionSyntax node)
         {
             var argumentSyntax = (ArgumentSyntax)node?.Parent;
```
>var argumentSyntax = (ArgumentSyntax)node?.Parent; [](start = 12, length = 50)

I don't think we ever expect the ```node``` to be null here and the code below doesn't expect ```argumentSyntax``` to be null, please remove the conditional access.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ExpressionVariableFinder.cs:
```diff
@@ -260,7 +260,7 @@ public override void VisitBinaryExpression(BinaryExpressionSyntax node)
             var argumentSyntax = (ArgumentSyntax)node?.Parent;
-            var argumentListSyntax = (BaseArgumentListSyntax)argumentSyntax?.Parent;
+            var argumentListSyntax = argumentSyntax.Parent as BaseArgumentListSyntax;
             var variable = MakeOutVariable(node, argumentListSyntax, _nodeToBind);
```
>var variable = MakeOutVariable(node, argumentListSyntax, _nodeToBind); [](start = 12, length = 70)

The ```MakeOutVariable``` name is now misleading, please rename so that the name wouldn't make people believe that the function is used only for legitimate Out Vars.

gafter replied:
Sure. I'll rename `_localsBuilder` too.

AlekseyTs commented:
>     <Field Name="Variable" Type="ExpressionSyntax">

Should the list of kinds within this element be adjusted?

---
Refers to: src/Compilers/CSharp/Portable/Syntax/Syntax.xml:2120 in c4d4991. [](commit_id = c4d4991d80eb3deae9349d1153fbc3467c7658b1, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -99,7 +99,8 @@ protected override ImmutableArray<LocalSymbol> BuildLocals()
                     break;
                 default:
-                    throw ExceptionUtilities.UnexpectedValue(declaration.Kind());
+                    ExpressionVariableFinder.FindExpressionVariables(this, locals, declaration);
```
>ExpressionVariableFinder.FindExpressionVariables(this, locals, declaration); [](start = 20, length = 76)

Could you please add a comment about the scenario when we get here and point me to the test(s) that covers this line of code?

gafter replied:
Sure. The tests would be `UnitTests.CodeGen.CodeGenDeconstructTests.MixedDeconstruction_06`, `UnitTests.CodeGen.CodeGenDeconstructTests.ForeachIntoExpression`, and `UnitTests.CodeGen.CodeGenDeconstructTests.MixedDeconstruction_05`. Comments in the next iteration.

---
In reply to: [91385737](https://github.com/dotnet/roslyn/pull/15548#discussion_r91385737) [](ancestors = 91385737)

AlekseyTs commented:
>                 // for error recovery, we parse `int (x, y) = e` as a deconstruction.

Should we keep the comment, is it still valid?

---
Refers to: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:9220 in 1003470. [](commit_id = 1003470c5f0de8925d3e1f465a59f549e2592eb2, deletion_comment = True)

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs:
```diff
@@ -102,6 +102,148 @@ private PatternSyntax ParsePattern()
+        // absence of the comma after the `D` means we don't treat the `D` as contributing to the
+        // disambiguation of the expression/type. More formally, ...
+        //
+        // If a sequence of tokens can be parsed(in context) as a* simple-name* (§7.6.3), *member-access* (§7.6.5),
```
>*member-access* [](start = 90, length = 15)

Consider removing the ```**``` annotations throughout this comment.

gafter replied:
I'd prefer to keep the markup from the specification.

---
In reply to: [91392143](https://github.com/dotnet/roslyn/pull/15548#discussion_r91392143) [](ancestors = 91392143)

AlekseyTs replied:
Is this ```a* simple-name*``` intended markup, note position of the first ```*```? There are several similar places throughout this comment. 

---
In reply to: [91453230](https://github.com/dotnet/roslyn/pull/15548#discussion_r91453230) [](ancestors = 91453230,91392143)

gafter replied:
That is what is in the draft spec. I agree it could be improved ever so slightly. If this is what is holding up your approval of the code review, I think we should hold this set of changes until I am back in the office in January.

---
In reply to: [91570733](https://github.com/dotnet/roslyn/pull/15548#discussion_r91570733) [](ancestors = 91570733,91453230,91392143)

jaredpar replied:
Let's not block on the markup here.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -670,9 +673,14 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
-                return result;
+                if ((object)this._type == null)
+                {
+                    diagnostics.Add(this.ForbiddenDiagnostic, Locations[0], Name);
```
>diagnostics.Add(this.ForbiddenDiagnostic, Locations[0], Name); [](start = 20, length = 62)

It is not clear why ```ForbiddenDiagnostic``` is an appropriate diagnostics to use in this case. Also, could you please point me to the test(s) covering this line?

gafter replied:
I don't know of a way to trigger this defensive code. I'll remove it and let the assertion in the caller fail if it occurs.

---
In reply to: [91395018](https://github.com/dotnet/roslyn/pull/15548#discussion_r91395018) [](ancestors = 91395018)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -670,9 +673,14 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+                {
+                    diagnostics.Add(this.ForbiddenDiagnostic, Locations[0], Name);
+                    SetType(new ExtendedErrorTypeSymbol(
+                        ((SourceModuleSymbol)this.ContainingModule).DeclaringCompilation, name: "var", arity: 0, errorInfo: null, variableUsedBeforeDeclaration: true));
```
>variableUsedBeforeDeclaration: true) [](start = 130, length = 36)

Why does it make sense to set ```variableUsedBeforeDeclaration``` to true here and what effect is this going to have?

gafter replied:
It wouldn't have any effect other than recording that the variable's type was required at a location before it was able to be inferred.

---
In reply to: [91395381](https://github.com/dotnet/roslyn/pull/15548#discussion_r91395381) [](ancestors = 91395381)

AlekseyTs commented:
Done with review pass. A number of active comments remains, I think I reactivated a several existing threads as well.

gafter commented on src/Compilers/CSharp/Portable/Binder/ExpressionVariableFinder.cs:
```diff
@@ -260,7 +260,7 @@ public override void VisitBinaryExpression(BinaryExpressionSyntax node)
         {
             var argumentSyntax = (ArgumentSyntax)node?.Parent;
-            var argumentListSyntax = (BaseArgumentListSyntax)argumentSyntax?.Parent;
+            var argumentListSyntax = argumentSyntax.Parent as BaseArgumentListSyntax;
```
There are five at the moment. See `OutVarTests.OutVarDeconstruction_02`, `OutVarTests.OutVarDeconstruction_03`, `OutVarTests.OutVarDeconstruction_04`, `PatternSwitchTests.TupleInPattern`, and `LocalFunctionTests.NoOperator` is also affected.

gafter commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -752,7 +760,13 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
-                Debug.Assert((object)this._type != null);
+                if ((object)this._type == null)
+                {
+                    diagnostics.Add(this.ForbiddenDiagnostic, Locations[0], Name);
```
Indeed, diagnostics from this method are always discarded. I'll remove it.

This is covered by `OutVarTests.OutVarDeconstruction_02`

There isn't a mechanism to "enforce" that we only get here for broken code, because we don't have the necessary compilation diagnostics in hand. Ideas for doing that are welcome.

---
In reply to: [91397639](https://github.com/dotnet/roslyn/pull/15548#discussion_r91397639) [](ancestors = 91397639,91395620)

AlekseyTs replied:
> There isn't a mechanism to "enforce" that we only get here for broken code, because we don't have the necessary compilation diagnostics in hand. Ideas for doing that are welcome.

I believe we should never get here for an Out Var or a pattern variable. It should be fairly easy to detect such cases by examining the syntax. Please add the detection and a throw for them.

---
In reply to: [91455715](https://github.com/dotnet/roslyn/pull/15548#discussion_r91455715) [](ancestors = 91455715,91397639,91395620)

gafter commented:
>     <Field Name="Variable" Type="ExpressionSyntax">

I will remove the list.

---
In reply to: [265570868](https://github.com/dotnet/roslyn/pull/15548#issuecomment-265570868) [](ancestors = 265570868)

---
Refers to: src/Compilers/CSharp/Portable/Syntax/Syntax.xml:2120 in c4d4991. [](commit_id = c4d4991d80eb3deae9349d1153fbc3467c7658b1, deletion_comment = False)

gafter commented:
>                 // for error recovery, we parse `int (x, y) = e` as a deconstruction.

Yes.

---
In reply to: [265577755](https://github.com/dotnet/roslyn/pull/15548#issuecomment-265577755) [](ancestors = 265577755)

---
Refers to: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:9220 in 1003470. [](commit_id = 1003470c5f0de8925d3e1f465a59f549e2592eb2, deletion_comment = True)

gafter commented:
git repository authentication is broken again, so I can't push to the repository. When I can I'll send the next iteration, which addresses all comments.

gafter commented:
>                 // for error recovery, we parse `int (x, y) = e` as a deconstruction.

More specifically, this is now the name of a kind of assignment expression in which the left-hand-side is of one of a couple of forms.

---
In reply to: [265670754](https://github.com/dotnet/roslyn/pull/15548#issuecomment-265670754) [](ancestors = 265670754,265577755)

---
Refers to: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:9220 in 1003470. [](commit_id = 1003470c5f0de8925d3e1f465a59f549e2592eb2, deletion_comment = True)

gafter commented:
All comments have been addressed. May I have a second code review, please?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -670,9 +673,7 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
                 }
 
-                TypeSymbol result = this._type;
-                Debug.Assert((object)result != null);
```
>Debug.Assert((object)result != null); [](start = 16, length = 37)

Please restore the original assert. The earlier we detect the problem the better.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:
```diff
@@ -5546,5 +5509,330 @@ static void Main()
+{
+    static void Main(string[] args)
+    {
+        foreach (M(out var x1) in new[] { 1, 2, 3 })
```
>foreach (M(out var x1) in new[] { 1, 2, 3 }) [](start = 8, length = 44)

Please add similar test for a ```var``` pattern variable.

gafter replied:
See the very next test  here.

---
In reply to: [91582314](https://github.com/dotnet/roslyn/pull/15548#discussion_r91582314) [](ancestors = 91582314)

AlekseyTs commented on src/EditorFeatures/CSharpTest/Formatting/Indentation/SmartIndenterTests.cs:
```diff
@@ -2570,7 +2570,7 @@ class C
         }
 
-        [WpfFact, Trait(Traits.Feature, Traits.Features.SmartIndent)]
+        [WpfFact(Skip = "test is no longer correct due to language changes"), Trait(Traits.Feature, Traits.Features.SmartIndent)]
```
>[WpfFact(Skip = "test is no longer correct due to language changes") [](start = 8, length = 68)

Do we have an issue tracking this?

gafter replied:
Tracking what? Language changes? The deletion of this test?

---
In reply to: [91584597](https://github.com/dotnet/roslyn/pull/15548#discussion_r91584597) [](ancestors = 91584597)

AlekseyTs replied:
If the test is skipped, a follow-up should follow. It can take many forms: remove, adjust, test new scenarios enabled by the change. Deciding what to do will be a part of resolving that issue, unless we can do that in context of this change.

---
In reply to: [91585635](https://github.com/dotnet/roslyn/pull/15548#discussion_r91585635) [](ancestors = 91585635,91584597)

jaredpar replied:
We should not be disabling tests in this manner.  If a test is disabled it needs a tracking issue.  If it's no longer valid then it should be deleted and we should make the appropriate team (IDE in this case) aware of the change.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -752,7 +753,11 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
-                Debug.Assert((object)this._type != null);
+                if ((object)this._type == null)
+                {
+                    SetType(_nodeBinder.CreateErrorType("var"));
```
>SetType(_nodeBinder.CreateErrorType("var")); [](start = 20, length = 44)

> There isn't a mechanism to "enforce" that we only get here for broken code, because we don't have the necessary compilation diagnostics in hand. Ideas for doing that are welcome.

I believe we should never get here for an Out Var or a pattern variable. It should be fairly easy to detect such cases by examining the syntax. Please add the detection and a throw for them.

AlekseyTs commented:
I am done with review pass. The remaining issues that I think should be addressed are:
- Restoring an original assert lost between iterations in SourceLocalSymbol.cs https://github.com/dotnet/roslyn/pull/15548#pullrequestreview-12087256
- Creating an issue tracking a follow-up for a test disabled in SmartIndenterTests.cs  https://github.com/dotnet/roslyn/pull/15548#pullrequestreview-12094564
- Adding an assert or a throw to catch unexpected type inference failure early in SourceLocalSymbol.cs https://github.com/dotnet/roslyn/pull/15548#pullrequestreview-12103761

AlekseyTs commented:
LGTM

jcouv commented:
@dotnet-bot test linux_debug_prtest please

// reason: error MSB4019: The imported project "/home/dotnet-bot/.nuget/packages/Microsoft.Net.RoslynDiagnostics/1.2.0-beta2/build/Microsoft.Net.RoslynDiagnostics.props" was not found. Confirm that the path in the declaration is correct, and that the file exists on disk.

jcouv commented:
@dotnet-bot test linux_debug_prtest please
